File: rpc/rpc-2018-01/alien.py
from itertools import cycle, islice

class Settlement:
    def __init__(self, h, r, t):
        self.h, self.r, self.t = h, r, t

    def is_night(self, x):
        if self.r < self.t:
            return x <= self.r or x >= self.t
        else:
            return x <= self.r and x >= self.t

    @property
    def night_hours(self):
        return cycle(map(self.is_night, range(self.h)))

def hours_until_darkness(settlements):
    longest_solar_day = max(s.h for s in settlements)
    president_term = 1825 * longest_solar_day
    hours_by_settlement_until_term = (islice(s.night_hours, president_term) for s in settlements)
    hours_by_day_until_term = transpose(hours_by_settlement_until_term)
    return find_index(all, hours_by_day_until_term)

def transpose(iterables):
    return zip(*iterables)

def find_index(pred, iterable):
    for i, item in enumerate(iterable):
        if pred(item):
            return i
    return -1

def read_settlement():
    return Settlement(*map(int, input().split(' ')))

if __name__ == '__main__':
    n = int(input())
    settlements = [read_settlement() for _ in range(n)]
    ans = hours_until_darkness(settlements)
    print("impossible" if ans == -1 else ans)

File: rpc/rpc-2018-01/Alien.hs
module AlienSunset where

import Control.Monad (replicateM)
import Data.Maybe    (listToMaybe)
import Data.List     (transpose, findIndex)

data Settlement
    = Settlement
    { solarDayLength :: Int
    , sunriseHour    :: Int
    , sunsetHour     :: Int
    }

isNightFor :: Settlement -> Int -> Bool
isNightFor (Settlement _ r t)
    | r < t     = \x -> x <= r || x >= t
    | otherwise = \x -> x <= r && x >= t

makeSolarDayHours :: Settlement -> [Bool]
makeSolarDayHours settlement = cycle (map (isNightFor settlement) hoursOfSolarDay)
    where
        hoursOfSolarDay = take (solarDayLength settlement) [0..]

hoursUntilDarkness :: [Settlement] -> Maybe Int
hoursUntilDarkness settlements = findIndex and hoursByDayUntilPresidentTerm
    where
        longestSolarDay                     = maximum (map solarDayLength settlements)
        presidentTerm                       = 1825 * longestSolarDay
        hoursBySettlement                   = map makeSolarDayHours settlements
        hoursBySettlementUntilPresidentTerm = map (take presidentTerm) hoursBySettlement
        hoursByDayUntilPresidentTerm        = transpose hoursBySettlementUntilPresidentTerm

readLnSpaced :: Read a => IO [a]
readLnSpaced = map read . words <$> getLine

readSettlement :: IO Settlement
readSettlement = do
    [h, r, t] <- readLnSpaced
    pure (Settlement h r t)

readProblemInput :: IO [Settlement]
readProblemInput = do
    n <- readLn
    replicateM n readSettlement

solveProblem :: [Settlement] -> String
solveProblem settlements =
    case hoursUntilDarkness settlements of
        Nothing -> "impossible"
        Just h  -> show h

runCase :: IO ()
runCase = putStrLn =<< solveProblem <$> readProblemInput

main = do
    print (hoursUntilDarkness [Settlement 24 7 19, Settlement 24 18 6])
    print (hoursUntilDarkness [Settlement 10 8 2, Settlement 15 5 10, Settlement 20 15 10])
    print (hoursUntilDarkness [Settlement 6 4 2, Settlement 12 7 5])
    print (hoursUntilDarkness [Settlement 10 5 6, Settlement 10 6 5])
